firmware/src/core/utils/TaskManager.cpp

```diff
@@ -0,0 +1,155 @@
+#include "TaskManager.h"
+#include "GlobalResources.h"
+#include "Utils.h"
+#include <HTTPClient.h>
+#include <memory>
+
+TaskManager *TaskManager::instance = nullptr;
+QueueHandle_t TaskManager::requestQueue = nullptr;
+QueueHandle_t TaskManager::responseQueue = nullptr;
+volatile uint32_t TaskManager::activeRequests = 0;
+volatile uint32_t TaskManager::maxConcurrentRequests = 0;
+int TaskManager::taskParamsCount = 0;
+
+TaskManager::TaskManager() {
+    if (!requestQueue) {
+        requestQueue = xQueueCreate(REQUEST_QUEUE_SIZE, REQUEST_QUEUE_ITEM_SIZE);
+    }
+    if (!responseQueue) {
+        responseQueue = xQueueCreate(RESPONSE_QUEUE_SIZE, RESPONSE_QUEUE_ITEM_SIZE);
+    }
+}
+
+TaskManager *TaskManager::getInstance() {
+    if (!instance) {
+        instance = new TaskManager();
+    }
+    return instance;
+}
+
+bool TaskManager::addTask(std::unique_ptr<Task> task) {
+    if (isUrlInQueue(task->url)) {
```

Comment:
Should we emit a message about suppressing duplicate requests?  _technically_ they might not be duplicate based on URL if they have different post-processing (but that's super edge-case).